File: common/src/main/java/com/codeloom/common/language/ValidLanguage.java
package com.codeloom.common.language;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import jakarta.validation.constraints.NotBlank;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@NotBlank
@Constraint(validatedBy = SubmissionLanguageValidator.class)
@Target(ElementType.FIELD)
@Retention(RetentionPolicy.RUNTIME)
public @interface ValidLanguage {
    String message() default "Invalid submission language";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: common/src/main/kotlin/com/codeloom/common/language/ValidLanguage.kt
package com.codeloom.common.language

import jakarta.validation.Constraint
import jakarta.validation.Payload
import jakarta.validation.constraints.NotBlank
import kotlin.reflect.KClass

@NotBlank
@Constraint(validatedBy = [SubmissionLanguageValidator::class])
@Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.RUNTIME)
annotation class ValidLanguage(
    val message: String = "Invalid submission language",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
)